src/Control/Algebra.hs

```diff
@@ -151,13 +157,13 @@ instance Algebra sig m => Algebra sig (Ap m) where
 -- > getAlt (mconcat [a, b, c, d])
 --
 -- @since 1.0.1.0
-instance Algebra sig m => Algebra sig (Alt m) where
-  alg = Alt . alg . handleCoercible
+deriving instance Algebra sig m => Algebra sig (Alt m)
 
 instance Algebra sig m => Algebra (Reader r :+: sig) (Reader.ReaderT r m) where
-  alg (L (Ask       k)) = Reader.ask >>= k
-  alg (L (Local f m k)) = Reader.local f m >>= k
-  alg (R other)         = Reader.ReaderT $ \ r -> alg (hmap (`Reader.runReaderT` r) other)
+  alg hom = \case
+    L (Ask       k) -> Reader.ask >>= hom . k
+    L (Local f m k) -> Reader.local f (hom m) >>= hom . k
+    R other         -> Reader.ReaderT $ \ r -> alg ((`Reader.runReaderT` r) . hom) other
```

Comment:
Algebras using `hmap` or `handleCoercible` can instead compose their handler onto the homomorphism passed to `alg` for the rest of the stack.